### PR TITLE
feat(mail): support large email attachments

### PR DESCRIPTION
**Pull request by chanthuang**
## Summary

Adds support for sending emails with attachments that exceed the 25MB EML size limit — up to 3GB per file. Oversized attachments are automatically uploaded to mail attachment storage and rendered as a download-card in the email body, matching the desktop client's behavior.

## What's new

**Large attachments, automatically handled**

- Pass any file path with `--attach` in `mail +send` / `+draft-create` / `+reply` / `+reply-all` / `+forward`. When the accumulated attachments would push the email past 25MB, the overflow files are uploaded as large attachments instead of embedded — no extra flag needed.
- Each large attachment is represented as a download-card in the email body (with filename, size, and a download link). External recipients without the Lark/Feishu client can still download via the link.
- Single-file cap: 3GB.

**Large attachment management in draft-edit**

- `mail +draft-edit` now treats large attachments as first-class siblings of normal attachments, with a consistent mental model:
  - Add: `{ op: "add_attachment", path: "..." }` — automatically routes to large-attachment upload when the file would overflow 25MB.
  - Remove: `{ op: "remove_attachment", target: { token: "..." } }` — by file token, same op used for normal attachments (which use `part_id` / `cid`).
  - Inspect: `--inspect` returns a new `large_attachments_summary` listing each large attachment's token, filename, and size.
- Body-edit ops (`set_body` / `set_reply_body`) automatically preserve signature block and attachment cards (normal + large), mirroring how normal attachment MIME parts survive body edits. Users delete signatures/attachments via the dedicated ops (`remove_signature`, `remove_attachment`), not by clearing the body.

**Reply/reply-all: strip large attachment card from quote**

- When replying or replying-all to a message that contains large attachment download cards, the card HTML is now stripped from the quoted block. Replies should not carry forward the original sender's large attachment references.

**Forward: skip invalid/deleted attachments automatically and print warning info**

- Forwarding a message with deleted or expired attachments (both normal and large) no longer blocks the operation. Invalid attachments are automatically removed — including their HTML download-card and header token for large attachments — and a warning is emitted listing the skipped files. Valid attachments proceed normally.
- `RemoveLargeFileItemFromHTML` now matches server-generated cards (which use `href?token=`) in addition to CLI-generated cards (`data-mail-token` attribute).

**Localization and branding**

- Card title and download text follow the CLI's language setting: Chinese uses "来自{飞书/Lark}邮箱的超大附件" / "下载", English uses "Large file from {Feishu/Lark} Mail" / "Download".
- Product name switches between Feishu/飞书 (domestic brand) and Lark (international brand).

## Documentation

Skill references updated under `skills/lark-mail/references/` for all affected compose shortcuts, and `lark-mail-draft-edit.md` documents the unified attachment model, the new `large_attachments_summary` field, and the body-edit preservation semantics.

## Test plan

- [ ] `mail +send` / `+reply` / `+reply-all` / `+forward` with a mix of small and large attachments
- [ ] `mail +draft-create` with signature + large attachment
- [ ] `mail +draft-edit --inspect` on a draft with large attachments — verify `large_attachments_summary` is populated
- [ ] `mail +draft-edit` with `add_attachment` (oversized) on existing drafts
- [ ] `mail +draft-edit` with `remove_attachment` by token
- [ ] `mail +draft-edit` with `set_body` / `set_reply_body` — verify signature, large attachment card, and quote block (for set_reply_body) are preserved
- [ ] Verify external recipients (non-Lark/Feishu clients) see the download-card and can retrieve the file via the download link
- [ ] Verify the card title/download text switches between zh/en and between Feishu/Lark brands
- [ ] `mail +reply` / `+reply-all` on a message with large attachment cards — verify the card is stripped from the quote
- [ ] `mail +forward` on a message with a deleted/expired large attachment — verify the attachment is skipped with a warning, HTML card removed, and the forward succeeds
- [ ] `mail +forward` on a message with valid attachments — verify no regression, all attachments preserved

### CODE REVIEW

File: shortcuts/mail/draft/large_attachment_parse.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+		if grand := container.Parent; grand != nil {
+			grand.RemoveChild(container)
+		}
+	}
+	var buf bytes.Buffer
+	if err := xhtml.Render(&buf, doc); err != nil {
+		return htmlBody, false
+	}
+	return buf.String(), true
+}
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/draft/large_attachment_parse.go]: Fixed in dad9902. Added `stripHTMLEnvelope()` to strip the `<html><head></head><body>...</body></html>` wrapper from `xhtml.Render` output, keeping the DOM-based manipulation intact while avoiding the fragment wrapping issue.

File: shortcuts/mail/mail_reply_all.go [hunk 2]
```diff
@@ -189,13 +190,12 @@ var MailReplyAll = common.Shortcut{
...
 			bld = bld.TextBody([]byte(bodyStr + quoted))
 		}
-		allFilePaths := append(append(splitByComma(attachFlag), inlineSpecFilePaths(inlineSpecs)...), autoResolvedPaths...)
-		if err := checkAttachmentSizeLimit(runtime.FileIO(), allFilePaths, 0); err != nil {
+		allInlinePaths := append(inlineSpecFilePaths(inlineSpecs), autoResolvedPaths...)
+		emlBase := estimateEMLBaseSize(runtime.FileIO(), int64(len(body)), allInlinePaths, 0)
+		bld, err = processLargeAttachments(ctx, runtime, bld, composedHTMLBody, splitByComma(attachFlag), emlBase, 0)
+		if err != nil {
 			return err
 		}
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/mail_reply_all.go:213]: Fixed in dad9902. Added a guard in `processLargeAttachments`: when `htmlBody == ""` (plain-text mode) and oversized attachments are detected, we now return a clear error instead of silently injecting HTML. The check is in the shared function so it covers all compose shortcuts (send, reply, reply-all, forward).

File: skills/lark-mail/references/lark-mail-send.md [hunk 3]
```diff
@@ -169,6 +169,7 @@ lark-cli mail user_mailbox.drafts cancel_scheduled_send --params '{"user_mailbox
 - 使用 EML 构建器生成完整 MIME 邮件并 base64url 编码后发送。
 - `--attach` 作为普通附件添加。相对路径。
 - `--inline` 接受 JSON 数组，每项需提供 `cid`（唯一标识符，可用随机十六进制字符串）和 `file_path`（相对路径），作为 inline part 嵌入邮件。
+- **超大附件**：当附件导致 EML 总大小（headers + body + inline images + attachments，base64 编码后）超过 25 MB 时，超出的文件自动通过 `medias/upload_*` API 上传到云端，邮件正文中插入与飞书客户端一致的下载卡片 HTML。单个文件上限 3 GB，总附件数量上限 100 个。
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply skills/lark-mail/references/lark-mail-send.md]: Fixed in dad9902. Updated the doc to match `MaxAttachmentCount = 250` in `limits.go`.

File: shortcuts/mail/large_attachment.go [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+	files := make([]attachmentFile, 0, len(paths))
+	for _, p := range paths {
+		if strings.TrimSpace(p) == "" {
+			continue
+		}
+		info, err := fio.Stat(p)
+		if err != nil {
+			return nil, fmt.Errorf("failed to stat attachment %s: %w", p, err)
+		}
+		files = append(files, attachmentFile{
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go:130]: **Issue:** Oversized attachments bypass the existing blocked-extension policy. `statAttachmentFiles` only stats the path, and once a file is classified as oversized it never reaches `emlbuilder.AddFileAttachment` or `draft.addAttachment`, which are the paths that call `filecheck.CheckBlockedExtension`. A `.exe`, `.js`, or `.ps1` file that is rejected as a normal attachment can therefore be uploaded and sent as a large attachment in compose and draft-edit flows.

**Severity:** 🚫 Must Fix

**Suggested Fix:** Run the shared blocked-extension check for every `--attach` / `add_attachment` path before classification or upload, ideally inside this helper so both compose and draft-edit paths are covered. Add regression coverage for an oversized blocked extension.

File: shortcuts/mail/large_attachment.go [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+	}
+
+	// Remove oversized ops from the patch (keep normal ones for draft.Apply).
+	oversizedPaths := make(map[string]bool, len(classified.Oversized))
+	for _, f := range classified.Oversized {
+		oversizedPaths[f.Path] = true
+	}
+	var filteredOps []draftpkg.PatchOp
+	for _, op := range patch.Ops {
+		if op.Op == "add_attachment" && oversizedPaths[op.Path] {
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go]: **Issue:** Filtering oversized `add_attachment` operations by `op.Path` drops valid duplicate attachments. If a draft-edit patch adds the same file path twice and only the later occurrence overflows the EML limit, `oversizedPaths[f.Path] = true` causes this loop to remove every operation with that path, including the occurrence classified as normal. The normal duplicate never reaches `draft.Apply`, so the resulting draft silently loses an attachment the user requested.

**Severity:** 🚨 Should Fix

**Suggested Fix:** Preserve the original patch indices through classification, or consume an ordered multiset/count keyed by path so only the exact occurrences classified as oversized are removed. Add a regression test where duplicate `add_attachment` paths split between normal and large attachment handling.

File: shortcuts/mail/large_attachment.go [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+	}
+	return out
+}
+
+// injectLargeAttachmentHTMLIntoSnapshot finds the HTML body part in the
+// snapshot and inserts the large attachment HTML before the quote block
+// (or appends to the end if no quote).
+func injectLargeAttachmentHTMLIntoSnapshot(snapshot *draftpkg.DraftSnapshot, largeHTML string) {
+	htmlPart := draftpkg.FindHTMLBodyPart(snapshot.Body)
+	if htmlPart == nil {
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go:741]: **Issue:** Adding an oversized attachment to a text/plain-only draft can succeed without inserting any download card. `preprocessLargeAttachmentsForDraftEdit` uploads the file before calling this helper, and when `FindHTMLBodyPart` returns nil while `snapshot.Body` is non-nil, this helper just returns. The caller still writes `X-Lms-Large-Attachment-Ids` and removes the `add_attachment` op, leaving the draft with a token association but no visible link for recipients to download the file.

**Severity:** 🚨 Should Fix

**Suggested Fix:** Detect the missing HTML body before uploading, or make this helper return an error and abort/convert the draft when the card cannot be inserted. Add a draft-edit regression test for adding an oversized attachment to a text/plain-only draft.

File: shortcuts/mail/large_attachment.go (hunk 4 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:130]: Fixed in 5deff56.

Moved `CheckBlockedExtension` into `statAttachmentFiles`, so all `--attach` / `add_attachment` paths (compose, draft-edit, forward) are validated before classification. Added regression test covering `.exe`, `.js`, `.ps1`, `.bat`.

File: shortcuts/mail/large_attachment.go (hunk 5 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go]: Fixed in 5deff56.

Replaced path-based `oversizedPaths` map with index-based `oversizedIndices` map. Added `SourceIndex` field to `attachmentFile`; draft-edit sets it from the original patch op index, so filtering precisely removes only the oversized ops without affecting same-path normal duplicates.

File: shortcuts/mail/large_attachment.go (hunk 6 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:741]: Fixed in 5deff56.

Added an HTML body preflight check in `preprocessLargeAttachmentsForDraftEdit` **before** uploading: when `FindHTMLBodyPart` returns nil, we now return an error immediately. This mirrors the existing guard in the compose path (`processLargeAttachments:428`) and prevents uploading files that would produce a draft with token association but no visible download card.

File: shortcuts/mail/mail_forward.go [hunk 7]
```diff
@@ -200,24 +209,116 @@ var MailForward = common.Shortcut{
...
+			return output.ErrValidation("attachment count %d exceeds the limit of %d", totalCount, MaxAttachmentCount)
+		}
+		allFiles = append(allFiles, userFiles...)
+		classified := classifyAttachments(allFiles, emlBase)
+
+		// Embed normal attachments.
+		origIdx := 0
+		for _, f := range classified.Normal {
+			if f.Path == "" {
+				bld = bld.AddAttachment(origAtts[origIdx].content, "application/octet-stream", origAtts[origIdx].filename)
```
Comment (infeng) [inline shortcuts/mail/mail_forward.go]: This drops the original MIME type for forwarded attachments that stay embedded in the EML. We already preserve `att.ContentType` when downloading the source attachment above, but this line hard-codes `application/octet-stream`, so forwarded PDFs/images/etc. lose their type unless they overflow into the large-attachment path. That changes preview/open behavior compared with the source message. Can we pass `origAtts[origIdx].contentType` here instead?

File: shortcuts/mail/large_attachment.go (hunk 5 above)
Comment (infeng) [reply shortcuts/mail/large_attachment.go]: Re-checked the current head (`5deff563`). This thread is still open from the implementation side: `attachmentFile` still only has `Path/FileName/Size`, and `preprocessLargeAttachmentsForDraftEdit()` still filters by path via `oversizedPaths[f.Path]` at `shortcuts/mail/large_attachment.go:612-620`. So if the same path is added twice and only one occurrence overflows, both ops still get dropped together. I do not see the claimed `SourceIndex` / index-based filtering in the current code, so I’m keeping this unresolved.

File: shortcuts/mail/large_attachment.go (hunk 6 above)
Comment (infeng) [reply shortcuts/mail/large_attachment.go:741]: Re-checked the current head (`5deff563`). I don’t see the preflight guard in the current code yet: `preprocessLargeAttachmentsForDraftEdit()` still uploads oversized files before any HTML-body check (`shortcuts/mail/large_attachment.go:551-559`), and `injectLargeAttachmentHTMLIntoSnapshot()` still silently returns when `FindHTMLBodyPart(snapshot.Body) == nil && snapshot.Body != nil` (`shortcuts/mail/large_attachment.go:658-665`). So a text/plain-only draft can still upload/remove the patch op without ever getting a visible download card. Keeping this thread open.

File: shortcuts/mail/large_attachment.go (hunk 4 above)
Comment (infeng) [reply shortcuts/mail/large_attachment.go:130]: Re-checked the current head (`5deff563`). This is still not fixed in code: `statAttachmentFiles()` at `shortcuts/mail/large_attachment.go:112-128` still only stats the file and never calls the shared blocked-extension check, so oversized `.exe` / `.js` / `.ps1` paths can still bypass the normal `AddFileAttachment` / `draft.addAttachment` guard and go straight into the large-attachment upload path. I also don’t see a regression test covering oversized blocked extensions under `shortcuts/mail/*test.go`. Keeping this thread open.

File: shortcuts/mail/mail_forward.go (hunk 7 above)
Comment (chanthuang) [reply shortcuts/mail/mail_forward.go]: Fixed in a7b9733. Changed to use `origAtts[origIdx].contentType` so normal forwarded attachments retain their original MIME type. The `application/octet-stream` override is now only applied in the large attachment upload path (temp file written for oversized originals).

File: shortcuts/mail/large_attachment.go (hunk 5 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go]: This was fixed in 5deff56. The `SourceIndex` field and index-based filtering are in the current HEAD — the review thread is anchored on the original diff lines which don't reflect the fix commit. See the [commit diff](https://github.com/larksuite/cli/commit/5deff5639c63115b2b06fbb4854ba2ca0e447f18) for the changes.

File: shortcuts/mail/large_attachment.go (hunk 6 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:741]: This was fixed in 5deff56. The HTML body preflight guard (`FindHTMLBodyPart` check before upload) is in the current HEAD. The review thread is anchored on the original diff lines. See the [commit diff](https://github.com/larksuite/cli/commit/5deff5639c63115b2b06fbb4854ba2ca0e447f18) for the changes.

File: shortcuts/mail/large_attachment.go (hunk 4 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:130]: This was fixed in 5deff56. `statAttachmentFiles()` now calls `filecheck.CheckBlockedExtension` before stat, and a regression test (`TestStatAttachmentFiles_BlockedExtension`) covers `.exe`/`.js`/`.ps1`/`.bat`. The review thread is anchored on the original diff lines. See the [commit diff](https://github.com/larksuite/cli/commit/5deff5639c63115b2b06fbb4854ba2ca0e447f18) for the changes.

File: shortcuts/mail/mail_forward.go [hunk 8]
```diff
@@ -200,24 +198,101 @@ var MailForward = common.Shortcut{
...
+		// Embed normal attachments.
+		origIdx := 0
+		for _, f := range classified.Normal {
+			if f.Path == "" {
+				bld = bld.AddAttachment(origAtts[origIdx].content, origAtts[origIdx].contentType, origAtts[origIdx].filename)
+				origIdx++
+			} else {
+				bld = bld.AddFileAttachment(f.Path)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/mail_forward.go:271]: Fixed in 31e0255. Replaced linear `origIdx` counter with `SourceIndex` field on `attachmentFile`. Each entry from `origAtts` now carries its original index, so `classifyAttachments` reordering no longer causes content mismatch in the Normal/Oversized loops.

File: shortcuts/mail/mail_forward.go [hunk 9]
```diff
@@ -8,8 +8,11 @@ import (
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
+	"os"
+	"path/filepath"
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/mail_forward.go]: Fixed in 31e0255. Replaced `fmt.Sprintf(".lark-cli-fwd-%s", filepath.Base(att.filename))` with `os.CreateTemp("", ".lark-cli-fwd-*")`. Temp files now go to the system temp directory with random names, avoiding CWD pollution and predictable path issues. Also added cleanup on write failure.

File: shortcuts/mail/mail_forward.go [hunk 10]
```diff
@@ -200,24 +209,116 @@ var MailForward = common.Shortcut{
...
+		for _, f := range userFiles {
+			if f.Size > MaxLargeAttachmentSize {
+				return output.ErrValidation("attachment %s (%.1f GB) exceeds the %.0f GB single file limit",
+					f.FileName, float64(f.Size)/1024/1024/1024, float64(MaxLargeAttachmentSize)/1024/1024/1024)
+			}
+		}
+		totalCount := len(origAtts) + len(userFiles)
+		if totalCount > MaxAttachmentCount {
+			return output.ErrValidation("attachment count %d exceeds the limit of %d", totalCount, MaxAttachmentCount)
+		}
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/mail_forward.go]: Fixed in 31e0255. Changed `totalCount := len(origAtts) + len(userFiles)` to `totalCount := len(origAtts) + len(largeAttIDs) + len(userFiles)` so that original large attachments (already present via `ForwardAttachments`) are counted toward the 250 limit.

File: shortcuts/mail/mail_forward.go [hunk 11]
```diff
@@ -207,17 +208,20 @@ var MailForward = common.Shortcut{
...
 		}
-		for _, path := range splitByComma(attachFlag) {
-			bld = bld.AddFileAttachment(path)
+		allInlinePaths := append(inlineSpecFilePaths(inlineSpecs), autoResolvedPaths...)
+		origAttEMLBytes := int64(0)
+		for _, att := range origAtts {
+			origAttEMLBytes += estimateBase64EMLSize(int64(len(att.content)))
+		}
+		emlBase := estimateEMLBaseSize(runtime.FileIO(), int64(len(body)), allInlinePaths, origAttEMLBytes)
+		bld, err = processLargeAttachments(ctx, runtime, bld, composedHTMLBody, splitByComma(attachFlag), emlBase, len(origAtts))
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/mail_forward.go]: Fixed in 8cec5bb. Two changes:
1. Body size now uses `len(composedHTMLBody) + len(composedTextBody)` instead of `len(body)`, so the forward quote/headers are included in the estimate.
2. `addInlineImagesToBuilder` now returns total downloaded bytes, passed as `extraBytes` to `estimateEMLBaseSize` so source inline images are counted.

Same fix applied to reply, reply-all, send, and draft-create.

File: shortcuts/mail/mail_reply.go [hunk 12]
```diff
@@ -175,13 +176,12 @@ var MailReply = common.Shortcut{
 		} else {
 			bld = bld.TextBody([]byte(bodyStr + quoted))
 		}
-		allFilePaths := append(append(splitByComma(attachFlag), inlineSpecFilePaths(inlineSpecs)...), autoResolvedPaths...)
-		if err := checkAttachmentSizeLimit(runtime.FileIO(), allFilePaths, 0); err != nil {
+		allInlinePaths := append(inlineSpecFilePaths(inlineSpecs), autoResolvedPaths...)
+		emlBase := estimateEMLBaseSize(runtime.FileIO(), int64(len(body)), allInlinePaths, 0)
+		bld, err = processLargeAttachments(ctx, runtime, bld, composedHTMLBody, splitByComma(attachFlag), emlBase, 0)
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply shortcuts/mail/mail_reply.go]: Fixed in 8cec5bb. Same root cause as the forward comment — `len(body)` only measured the raw `--body` flag, not the composed body with quoted reply and signature. Now uses `len(composedHTMLBody) + len(composedTextBody)`. Source inline images are also counted via `addInlineImagesToBuilder` returning total downloaded bytes.

**Review by infeng — DISMISSED** [review]
Re-checked the latest head. All previously raised review items are addressed, local `go test ./shortcuts/mail/...` passes, and there are no unresolved review threads on the PR.

File: skills/lark-mail/references/lark-mail-send.md [hunk 13]
```diff
@@ -68,7 +68,7 @@ lark-cli mail +send --to alice@example.com --subject '测试' --body '<p>test</p
 | `--cc <emails>` | 否 | 抄送邮箱，多个用逗号分隔 |
 | `--bcc <emails>` | 否 | 密送邮箱，多个用逗号分隔 |
 | `--plain-text` | 否 | 强制纯文本模式，忽略 HTML 自动检测。不可与 `--inline` 同时使用 |
-| `--attach <paths>` | 否 | 附件文件路径，多个用逗号分隔。相对路径 |
+| `--attach <paths>` | 否 | 附件文件路径，多个用逗号分隔。相对路径。当附件导致 EML 总大小超过 25 MB 时，超出部分自动上传为超大附件（通过下载链接卡片呈现），单个文件上限 3 GB |
```
[the remark this answers is not in the document]
Comment (chanthuang) [reply skills/lark-mail/references/lark-mail-send.md]: This is no longer an issue — plain-text mode now supports large attachments as of commit 4859f7e.

File: shortcuts/mail/helpers.go [hunk 14]
```diff
@@ -2098,14 +2081,9 @@ func validateComposeInlineAndAttachments(fio fileio.FileIO, attachFlag, inlineFl
...
-	// dry-run and reply/forward can catch local errors before Execute.
-	// Auto-resolved local images are only known at Execute time, so Execute
-	// performs a second, complete size check that includes them.
-	inlineSpecs, err := parseInlineSpecs(inlineFlag)
-	if err != nil {
-		return err
-	}
-	allFiles := append(splitByComma(attachFlag), inlineSpecFilePaths(inlineSpecs)...)
-	return checkAttachmentSizeLimit(fio, allFiles, 0)
+	// Validate inline specs format early (before Execute).
```
Comment (infeng) [inline shortcuts/mail/helpers.go]: This regresses `--dry-run` validation for explicit local files. `validateComposeInlineAndAttachments()` now only parses the inline JSON, but the framework runs `Validate()` before `--dry-run` and skips `Execute()` afterward (`shortcuts/common/runner.go:637-644`). That means `mail +send/+reply/+forward/+draft-create --attach ./missing.txt --dry-run` now prints a request instead of surfacing the local path / blocked-extension error. Consider restoring lightweight preflight checks here for explicit `--attach` / `--inline` file paths, while still leaving the large-attachment size classification in `Execute()`.

File: shortcuts/mail/large_attachment.go [hunk 15]
```diff
@@ -0,0 +1,883 @@
...
+}
+
+// classifyAttachments splits files into normal (embed in EML) and oversized
+// (upload separately as large attachments).
+//
+// The decision is based on the estimated total EML size: headers + body +
+// inline images + attachments, all base64-encoded. Files are processed in
+// the user-specified order. Once a file would push the EML over MaxEMLSize,
+// it and all subsequent files are classified as oversized.
+func classifyAttachments(files []attachmentFile, emlBaseSize int64) classifiedAttachments {
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go:92]: `classifyAttachments()` only budgets against the current EML baseline. It never reserves space for the replacement payload that gets added later for `Oversized` files (`buildLargeAttachmentHTML` / `buildLargeAttachmentPlainText` plus the `X-Lms-Large-Attachment-Ids` header). On a message that is already close to `MaxEMLSize`, one file can be moved to `Oversized`, but the injected card/text/header can still push the final raw EML back over 25 MB and make `BuildBase64URL()` fail. This probably needs either a per-large-attachment overhead in the classifier or a second pass after injection.

File: shortcuts/mail/helpers.go (hunk 14 above)
Comment (chanthuang) [reply shortcuts/mail/helpers.go]: Fixed in cd8dada. Restored `fio` and `attachFlag` parameters in `validateComposeInlineAndAttachments` — it now calls `statAttachmentFiles` to verify file existence and blocked-extension checks before Execute. Size classification remains in `processLargeAttachments` during Execute. All 5 compose shortcuts (send, reply, reply-all, forward, draft-create) updated.

File: shortcuts/mail/large_attachment.go (hunk 15 above)
Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:92]: Fixed in cd8dada. Added `largeAttOverheadPerFile = 3KB` constant in `classifyAttachments` — each oversized file now reserves 3KB for the HTML card / plain-text block and header entry that gets injected after classification. This prevents borderline cases where the injected card pushes the final EML over 25MB.

File: shortcuts/mail/large_attachment.go [hunk 16]
```diff
@@ -0,0 +1,890 @@
...
+// card / plain-text block and header entry that will be injected later.
+func classifyAttachments(files []attachmentFile, emlBaseSize int64) classifiedAttachments {
+	const largeAttOverheadPerFile int64 = 3 * 1024
+
+	var result classifiedAttachments
+	accumulated := emlBaseSize
+	overflow := false
+
+	for _, f := range files {
+		if overflow {
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go:98]: Re-checked the current head (`cd8dada`). The new `largeAttOverheadPerFile` reserve still is not part of the overflow decision itself. In the boundary case where `accumulated` is just under 25 MB, a file can be classified as `Oversized` because `accumulated + cost > limit`, and then the code adds the 3 KB reserve afterward (`accumulated += largeAttOverheadPerFile`) without ever re-checking whether that reserve alone already pushes the final raw EML over `MaxEMLSize`. So `BuildBase64URL()` can still fail on the first oversized file. This likely needs the reserve to participate in the classification threshold, or a final post-classification/pass-to-build check that can force one more attachment out of the embedded set.

Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:98]: Agreed — the overhead reserve was addressing a boundary case that's practically impossible to trigger (Normal attachments would need to fill to within a few KB of 25MB AND have multiple oversized files). Removed the reserve in 57be277 to keep the classifier simple. The dry-run preflight fix from cd8dada is retained.

File: shortcuts/mail/large_attachment.go [hunk 17]
```diff
@@ -0,0 +1,883 @@
...
+			result.Oversized = append(result.Oversized, f)
+			continue
+		}
+		cost := estimateBase64EMLSize(f.Size)
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go:102]: Re-checked the current head (`57be277`). This change removes the only compensation for the extra payload introduced by large attachments, so the original correctness gap is back: `classifyAttachments()` still decides overflow using only the pre-upload EML estimate, but the final message is larger than that estimate because each oversized file also adds body content (HTML card or plain-text block) plus `X-Lms-Large-Attachment-Ids` header bytes after classification. That means a borderline message can still classify one file as `Oversized`, upload it successfully, and then fail later in `BuildBase64URL()` because the injected download-link payload was never budgeted. Even if uncommon, this is still a data-dependent false-negative rather than an impossible path, so I would keep a correctness guard here (reserve in the threshold, or a second pass/final reclassification check).

Comment (chanthuang) [reply shortcuts/mail/large_attachment.go:102]: Acknowledged. We're choosing not to handle this boundary case for now. It requires Normal attachments to fill to within a few KB of 25 MB AND have oversized files whose injected cards push it over — a combination that's extremely unlikely in practice. The added complexity (threshold adjustment or second-pass reclassification) isn't justified for this edge case. If it surfaces in real usage we can revisit.

Comment (infeng) [inline shortcuts/mail/large_attachment.go:102]: Re-checked the current head (`57be277`). I still don’t think this can be dismissed as requiring “multiple oversized files” or as a purely theoretical setup: a single oversized file is enough if the embedded portion of the message is already near the 25 MB limit, because `classifyAttachments()` still never budgets for the download-link payload that gets appended after the file is moved out of the MIME attachments. In that case the command can accept the input, upload the file, and then fail only at `BuildBase64URL()` with an EML-over-limit error. That is still a real correctness bug in the acceptance path, so I would keep this as should-fix rather than treating it as won’t-fix.

**Comment by infeng** [comment]
Re-checked the current head (`c478a63`). I don't have any additional must-fix / should-fix findings beyond the previously discussed large-attachment classification edge case that you've explicitly decided not to address in this PR. The latest push is test-only formatting (`mail_draft_create_test.go`), and `go test ./shortcuts/mail/...` still passes on this head.

Residual risk remains the same: the borderline EML-size acceptance path discussed in the thread at `shortcuts/mail/large_attachment.go` is still open by choice. Other than that, I don't see new blocking issues on the current head.

**Review by infeng — DISMISSED** [review]

File: shortcuts/mail/large_attachment.go [hunk 18]
```diff
@@ -0,0 +1,859 @@
...
+
+	results := make([]largeAttachmentResult, 0, len(files))
+	for _, f := range files {
+		fmt.Fprintf(runtime.IO().ErrOut, "Uploading large attachment: %s (%s)\n", f.FileName, common.FormatSize(f.Size))
+
+		var (
+			fileToken string
+			err       error
+		)
+		if f.Data != nil {
```
Comment (infeng) [inline shortcuts/mail/large_attachment.go:158]: Re-checked the current head (`912c8a3`). This refactor appears to re-introduce the original `>20MB forwarded attachment` regression that this PR had already fixed. In the new `f.Data != nil` branch, `uploadLargeAttachments()` always calls `UploadDriveMediaAll(...)` regardless of size, but `upload_all` is only the single-part path and the repo constant is still `MaxDriveMediaUploadSinglePartSize = 20MB`. For oversized original attachments forwarded from the source message, we now no longer fall back to multipart upload, so a `Path == "" && Size > 20MB` attachment can fail again even though the previous temp-file/multipart path handled it. I would keep the old multipart-capable behavior here (or add a reader-based multipart path) instead of routing every in-memory attachment through `upload_all`.

**Review by infeng — APPROVED** [review]
